File: src/main/java/cn/chahuyun/economy/entity/fish/FishPond.java
package cn.chahuyun.economy.entity.fish;

import cn.chahuyun.economy.plugin.FishManager;
import cn.chahuyun.economy.utils.EconomyUtil;
import cn.chahuyun.hibernateplus.HibernateFactory;
import cn.hutool.core.builder.EqualsBuilder;
import cn.hutool.core.builder.HashCodeBuilder;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * 鱼塘
 *
 * @author Moyuyanli
 * @date 2022/12/8 14:37
 */
@Entity(name = "FishPond")
@Table
@Getter
@Setter
public class FishPond implements Serializable {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private int id;
    /**
     * 鱼塘code<p>
     * 群鱼塘 g[群号]<p>
     * 私人鱼塘 [群号]-[玩家qq]<p>
     * 私人全局鱼塘 [玩家qq]<p>
     */
    private String code;
    /**
     * 鱼塘管理者
     */
    private long admin;
    /**
     * 鱼塘类型<p>
     * 1-群鱼塘<p>
     * 2-私人鱼塘<p>
     * 3-全局鱼塘<p>
     */
    private int pondType;
    /**
     * 鱼塘名称
     */
    private String name;
    /**
     * 鱼塘描述
     */
    private String description;
    /**
     * 鱼塘等级
     */
    private int pondLevel;
    /**
     * 限制最低进入等级
     */
    private int minLevel;
    /**
     * 鱼塘钓的鱼出售回扣金额
     * 0.00-0.10
     */
    private double rebate;
    /**
     * 总钓鱼次数
     */
    private int number;

    @OneToMany(targetEntity = Fish.class, mappedBy = "id", fetch = FetchType.EAGER)
    private List<Fish> fishList;

    public FishPond() {
    }

    public FishPond(int pondType, long group, long admin, String name, String description) {
        if (pondType == 1) {
            this.code = "g-" + group;
        } else if (pondType == 2) {
            this.code = "g-" + group + "-" + admin;
        } else {
            this.code = String.valueOf(admin);
        }
        this.admin = admin;
        this.name = name;
        this.description = description;
        if (pondType == 1) {
            this.pondLevel = 6;
        } else {
            this.pondLevel = 1;
        }
        this.pondType = pondType;
        this.minLevel = 0;
        this.rebate = 0.05;
        this.number = 0;
    }

    public Long getGroup() {
        Pattern compile = Pattern.compile("g-(\\d+)");
        Matcher matcher = compile.matcher(this.code);
        if (matcher.find()) {
            return Long.parseLong(matcher.group());
        }
        return 0L;
    }

    /**
     * 获取鱼塘的经济
     *
     * @return double
     * @author Moyuyanli
     * @date 2022/12/12 9:56
     */
    public double getFishPondMoney() {
        return EconomyUtil.getMoneyFromPluginBankForId(getCode(), getDescription());
    }

    /**
     * 获取池塘的鱼
     *
     * @return 池塘的鱼
     */
    public List<Fish> getFishList() {
        if (this.fishList == null) {
            return new ArrayList<>();
        }
        return fishList;
    }

    /**
     * 获取池塘的鱼
     *
     * @param level 鱼的等级
     * @return 池塘的鱼
     */
    public List<Fish> getFishList(int level) {
        if (pondType == 1) {
            return FishManager.getLevelFishList(level);
        }
        if (this.fishList == null) {
            return new ArrayList<>();
        }
        return fishList;
    }

    /**
     * 添加一次钓鱼次数
     */
    public void addNumber() {
        this.number++;
        HibernateFactory.merge(this);
    }

    /**
     * 保存
     */
    public FishPond save() {
        return HibernateFactory.merge(this);
    }


    public int getId() {
        return id;
    }

    public FishPond setId(int id) {
        this.id = id;
        return this;
    }

    public String getCode() {
        return code;
    }

    public FishPond setCode(String code) {
        this.code = code;
        return this;
    }

    @Override
    public boolean equals(Object object) {
        if (this == object) return true;

        if (object == null || getClass() != object.getClass()) return false;

        FishPond fishPond = (FishPond) object;

        return new EqualsBuilder().append(code, fishPond.code).isEquals();
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder(17, 37).append(code).toHashCode();
    }
}

File: src/main/kotlin/cn/chahuyun/economy/entity/fish/FishPond.kt
package cn.chahuyun.economy.entity.fish

import cn.chahuyun.economy.plugin.FishManager
import cn.chahuyun.economy.utils.EconomyUtil
import cn.chahuyun.hibernateplus.HibernateFactory
import jakarta.persistence.*
import java.io.Serializable
import java.util.regex.Pattern

/**
 * 鱼塘
 *
 * @author Moyuyanli
 * @date 2022/12/8 14:37
 */
@Entity(name = "FishPond")
@Table
class FishPond(
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    var id: Int = 0,

    /**
     * 鱼塘code
     * 群鱼塘 g[群号]
     * 私人鱼塘 [群号]-[玩家qq]
     * 私人全局鱼塘 [玩家qq]
     */
    var code: String = "",

    /**
     * 鱼塘管理者
     */
    var admin: Long = 0,

    /**
     * 鱼塘类型
     * 1-群鱼塘
     * 2-私人鱼塘
     * 3-全局鱼塘
     */
    var pondType: Int = 0,

    /**
     * 鱼塘名称
     */
    var name: String? = null,

    /**
     * 鱼塘描述
     */
    var description: String? = null,

    /**
     * 鱼塘等级
     */
    var pondLevel: Int = 0,

    /**
     * 限制最低进入等级
     */
    var minLevel: Int = 0,

    /**
     * 鱼塘钓的鱼出售回扣金额
     * 0.00-0.10
     */
    var rebate: Double = 0.05,

    /**
     * 总钓鱼次数
     */
    var number: Int = 0,

    @OneToMany(targetEntity = Fish::class, fetch = FetchType.EAGER)
    @JoinColumn(name = "pond_id")
    var fishList: MutableList<Fish>? = null
) : Serializable {

    constructor(pondType: Int, group: Long, admin: Long, name: String?, description: String?) : this(
        code = when (pondType) {
            1 -> "g-$group"
            2 -> "g-$group-$admin"
            else -> admin.toString()
        },
        admin = admin,
        name = name,
        description = description,
        pondLevel = if (pondType == 1) 6 else 1,
        pondType = pondType,
        minLevel = 0,
        rebate = 0.05,
        number = 0
    )

    val group: Long
        get() {
            val matcher = Pattern.compile("g-(\\d+)").matcher(code)
            return if (matcher.find()) matcher.group(1).toLong() else 0L
        }

    /**
     * 获取鱼塘的经济
     */
    fun getFishPondMoney(): Double {
        return EconomyUtil.getMoneyFromPluginBankForId(code, description ?: "")
    }

    /**
     * 获取池塘的鱼
     */
    fun getFishList(level: Int): List<Fish> {
        if (pondType == 1) {
            return FishManager.getLevelFishList(level)
        }
        return fishList ?: mutableListOf()
    }

    /**
     * 添加一次钓鱼次数
     */
    fun addNumber() {
        this.number++
        HibernateFactory.merge(this)
    }

    /**
     * 保存
     */
    fun save(): FishPond {
        return HibernateFactory.merge(this)!!
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is FishPond) return false
        return code == other.code
    }

    override fun hashCode(): Int {
        return code.hashCode()
    }
}
